### PR TITLE
Add shims for file handling

### PR DESCRIPTION
**Pull request by christianpoveda**
This adds the bare minimum to be able to do `File::open` and `File::read`. I also need some feedback about how to handle certain things

### CODE REVIEW

File: src/shims/io.rs [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+            throw_unsup_format!("`open` not available when isolation is enabled")
+        }
+
+        let flag = this.read_scalar(flag_op)?.to_i32()?;
+
+        if flag
+            != this
+                .eval_path_scalar(&["libc", "O_RDONLY"])?
+                .unwrap()
+                .to_i32()?
```
Comment (christianpoveda) [inline src/shims/io.rs]: I was thinking about writing a `FileHandler::init()` function to populate a dictionary with all the `libc` constants used here so we dont have to do this everywhere. Opinions?

File: src/shims/io.rs [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+        path_op: OpTy<'tcx, Tag>,
+        flag_op: OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        if !this.machine.communicate {
+            throw_unsup_format!("`open` not available when isolation is enabled")
+        }
+
+        let flag = this.read_scalar(flag_op)?.to_i32()?;
```
Comment (christianpoveda) [inline src/shims/io.rs]: Which flags should we implement? I think that implementing `O_RDONLY`, `O_WRONLY` and `O_RDWR` would be enough

File: src/shims/io.rs [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+        let cmd = this.read_scalar(cmd_op)?.to_i32()?;
+
+        if cmd
+            == this
+                .eval_path_scalar(&["libc", "F_SETFD"])?
+                .unwrap()
+                .to_i32()?
+        {
+            let flag = this.read_scalar(arg_op.unwrap())?.to_i32()?;
+            this.machine.file_handler.flags.insert(fd, flag);
```
Comment (christianpoveda) [inline src/shims/io.rs]: How should I handle the flag changes for a file that is already opened? What should happen if someone decides to enable the `O_CREAT` flag for a file using `fcntl`?

File: tests/hello.txt [hunk 4]
```diff
@@ -0,0 +1 @@
+Hello, World!
```
Comment (christianpoveda) [inline tests/hello.txt]: Where should I put this file?

File: src/shims/io.rs (hunk 3 above)
Comment (oli-obk) [reply src/shims/io.rs]: idk, I'd be fine with erroring as unsupported until someone hits a use-case. Or do whatever the libc docs state

File: src/shims/io.rs (hunk 2 above)
Comment (oli-obk) [reply src/shims/io.rs]: Whatever is needed to get a few libstd examples to work

File: src/shims/io.rs (hunk 1 above)
Comment (oli-obk) [reply src/shims/io.rs]: idk, If there's enough of these to warrant special code I'd rather have a `eval_libc_i32` function or sth.

File: tests/hello.txt (hunk 4 above)
Comment (oli-obk) [reply tests/hello.txt]: *ideally* nowhere :D but I'm not sure if 

```rust
let dir = TempDir::new("my_directory_prefix").unwrap();
std::fs::write(dir.join("hello.txt"), "Hello, World!").unwrap();
assert_eq!(std::fs::read(dir.join("hello.txt")).unwrap(), "Hello, World!");
```

is already possible today

File: src/shims/io.rs (hunk 1 above)
Comment (christianpoveda) [reply src/shims/io.rs]: Ok I'll add it to `src/helpers.rs`

File: tests/hello.txt (hunk 4 above)
Comment (christianpoveda) [reply tests/hello.txt]: Jmmm... that would mean adding tempdir as a dev dependency, is that ok?

Comment (oli-obk) [reply tests/hello.txt]: yea, let's not do that :)

File: src/shims/io.rs (hunk 2 above)
Comment (christianpoveda) [reply src/shims/io.rs]: I'm going to work only with the `O_RDONLY` flag in this PR to make it easier to review

File: tests/hello.txt (hunk 4 above)
Comment (christianpoveda) [reply tests/hello.txt]: I'm going to keep that file temporarily until the `(mk|rm)dir` shims are implemented. I'll add a fixme somewhere.

File: src/shims/io.rs (hunk 3 above)
Comment (christianpoveda) [reply src/shims/io.rs]: Well that's the thing, when someone implements  `execve`  they would need to take care of this flag but at the moment nothing uses `execve`.

File: src/shims/io.rs [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+use std::io::Read;
+
+use crate::stacked_borrows::Tag;
+use crate::*;
+
+#[derive(Default)]
+pub struct FileHandler {
+    files: HashMap<i32, File>,
+    flags: HashMap<i32, i32>,
+    low: i32,
```
Comment (oli-obk) [inline src/shims/io.rs]: i think 0, 1, and 2 need to be kept reserved, we already support writing to stdout and stderr.

Comment (christianpoveda) [reply src/shims/io.rs]: Oh shoo, that's true

File: src/shims/io.rs [hunk 6]
```diff
@@ -0,0 +1,170 @@
+use std::collections::HashMap;
+use std::fs::File;
+use std::io::Read;
+
+use crate::stacked_borrows::Tag;
+use crate::*;
+
+pub struct FileHandler {
+    files: HashMap<i32, File>,
```
Comment (oli-obk) [inline src/shims/io.rs]: `files` and `flags` always have the same entries, right? If so, create a `FileHandle` struct that has the `File` field and a `flag` field and only use one `HashMap`

Comment (christianpoveda) [reply src/shims/io.rs]: Yes, they have the same entries.

File: src/shims/io.rs [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+            if let Some(old_flag) = this.machine.file_handler.flags.get_mut(&fd) {
+                if flag ^ *old_flag == fd_cloexec {
+                    *old_flag = flag;
+                } else {
+                    throw_unsup_format!("Unsupported arg {:#x} for `F_SETFD`", flag);
+                }
+            }
+            Ok(0)
+        } else if cmd == this.eval_libc_i32("F_GETFD")? {
+            if let Some(flag) = this.machine.file_handler.flags.get(&fd) {
```
Comment (oli-obk) [inline src/shims/io.rs]: This get-or-EBADF business should get its own method (after the `flags`/`files` merger this will be usable everywhere I think)

File: src/shims/io.rs [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+        let this = self.eval_context_mut();
+
+        if !this.machine.communicate {
+            throw_unsup_format!("`open` not available when isolation is enabled")
+        }
+
+        let fd = this.read_scalar(fd_op)?.to_i32()?;
+
+        if let Some(file) = this.machine.file_handler.files.remove(&fd) {
+            match file.sync_all() {
```
Comment (oli-obk) [inline src/shims/io.rs]: a method consuming an `io::Result<T>` and spitting out a `InterpResult<'tcx, T>` would probably reduce a lot of repetition of

```
Err(e) => {
                    this.machine.last_error = e.raw_os_error().unwrap() as u32;
                    Ok(-1)
                }
```

File: src/shims/io.rs (hunk 7 above)
Comment (oli-obk) [reply src/shims/io.rs]: such a method could even take a `OpTy<'tcx, Tag>` for the fd directly

Comment (christianpoveda) [reply src/shims/io.rs]: You mean as a method for the `FileHandler`, right?

File: src/shims/io.rs [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+        }
+
+        let tcx = &{ this.tcx.tcx };
+
+        let fd = this.read_scalar(fd_op)?.to_i32()?;
+        let buf = this.force_ptr(this.read_scalar(buf_op)?.not_undef()?)?;
+        let count = this.read_scalar(count_op)?.to_usize(&*this.tcx)?;
+
+        if let Some(file) = this.machine.file_handler.files.get_mut(&fd) {
+            let mut bytes = vec![0; count as usize];
```
Comment (oli-obk) [inline src/shims/io.rs]: I'm not sure if that works out, but I think you can use https://doc.rust-lang.org/nightly/nightly-rustc/rustc_mir/interpret/struct.Allocation.html#method.get_bytes_mut to get a `&mut [u8]` to the destination directly and then read directly into the destination allocation.

that would probably require removing the `FileHandle` from the `files` map and inserting it back in later, because otherwise you borrow `this` via `memory_mut()` and via `this.machine`

File: src/shims/io.rs (hunk 7 above)
Comment (christianpoveda) [reply src/shims/io.rs]: > such a method could even take a `OpTy<'tcx, Tag>` for the fd directly

I think that just taking the `i32` is better because taking the `OpTy` would mean evaluating it several times.

Comment (oli-obk) [reply src/shims/io.rs]: oh, yea in that case just use the i32

Comment (oli-obk) [reply src/shims/io.rs]: > You mean as a method for the FileHandler, right?

yes

Comment (christianpoveda) [reply src/shims/io.rs]: I think this method and the `consume_result` one shouldn't be public at all, what do you think?

File: src/shims/io.rs [hunk 10]
```diff
@@ -48,21 +48,16 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
...
-                Ok(fh.low)
-            }
-
-            Err(e) => {
-                this.machine.last_error = e.raw_os_error().unwrap() as u32;
-                Ok(-1)
-            }
-        }
+            .map_err(|_| err_unsup_format!("{:?} is not a valid utf-8 string", path_bytes))?
+            .to_owned();
```
Comment (oli-obk) [inline src/shims/io.rs]: ŵhy the `to_owned`?

File: src/shims/io.rs [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+        let this = self.eval_context_mut();
+        if let Some(handle) = this.machine.file_handler.handles.remove(&fd) {
+            f(handle, this)
+        } else {
+            this.machine.last_error = this.eval_libc_i32("EBADF")? as u32;
+            Ok(t)
+        }
+    }
+
+    fn consume_result<T>(&mut self, result: std::io::Result<T>, t: T) -> InterpResult<'tcx, T> {
```
Comment (oli-obk) [inline src/shims/io.rs]: is anything but `-1: i32` ever used for `t: T`? If not, remove the argument

Comment (oli-obk) [reply src/shims/io.rs]: same for the above functions

Comment (christianpoveda) [reply src/shims/io.rs]: well the thing is that in some cases `-1i64` is returned. I  didn't manage to return both in the same function without passing it as an argument

Comment (christianpoveda) [reply src/shims/io.rs]: well the thing is that in some cases -1i64 is returned. I didn't manage to return both in the same function without passing it as an argument

File: src/shims/io.rs (hunk 10 above)
Comment (christianpoveda) [reply src/shims/io.rs]: Probably it was some residue of a borrowing issue I had earlier. Will fix

File: src/shims/io.rs [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+        let mut bytes = vec![0; count as usize];
+
+        let read_result = if let Some(FileHandle { file, ..}) = this.machine.file_handler.handles.get_mut(&fd) {
+            file.read(&mut bytes).map(|bytes| bytes as i64)
+        } else {
+            this.machine.last_error = this.eval_libc_i32("EBADF")? as u32;
+            return Ok(-1);
+        };
+
+        let read_bytes = this.consume_result::<i64>(read_result, -1)?;
```
Comment (christianpoveda) [inline src/shims/io.rs]: @oli-obk, for example here I could just use `-1i32` and then cast it to an `i64` however I'm not sure if it is ok to just cast `read_result` to an `i32` instead.

File: src/shims/io.rs (hunk 11 above)
Comment (oli-obk) [reply src/shims/io.rs]: I would have thought all the C-stuff to use `c_int` which is `i32`. Where is `i64` needed?

Comment (christianpoveda) [reply src/shims/io.rs]: for `read` (and `write` eventually) but I was able to fix it using `From<i32>` :P

File: src/shims/io.rs [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+            |handle, this| this.consume_result(handle.file.sync_all().map(|_| 0i32)),
+        )
+    }
+
+    fn read(
+        &mut self,
+        fd_op: OpTy<'tcx, Tag>,
+        buf_op: OpTy<'tcx, Tag>,
+        count_op: OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, i64> {
```
[the remark this answers is not in the document]
Comment (oli-obk) [reply src/shims/io.rs]: heh, `isize` should not be used, we're in the interpreter, so `i64` is totally alright. But this should be documented ;) Can you throw around some docs on all the handler functions explaining their arguments and return type?

[the remark this answers is not in the document]
Comment (christianpoveda) [reply src/shims/io.rs]: Sure let me sprinkle some docs here and there

File: src/shims/foreign_items.rs [hunk 14]
```diff
@@ -929,6 +949,14 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         }
         return Ok(None);
     }
+
+    fn eval_libc_i32(&mut self, name: &str) -> InterpResult<'tcx, i32> {
```
Comment (RalfJung) [inline src/shims/foreign_items.rs]: Please add a doc comment. I had to read the code to understand what this does.^^

Also, seems like both of these `eval` functions might have a better home in `helpers.rs`?

File: src/lib.rs [hunk 15]
```diff
@@ -34,6 +34,7 @@ pub use crate::shims::intrinsics::EvalContextExt as IntrinsicsEvalContextExt;
 pub use crate::shims::tls::{EvalContextExt as TlsEvalContextExt, TlsData};
 pub use crate::shims::dlsym::{Dlsym, EvalContextExt as DlsymEvalContextExt};
 pub use crate::shims::env::{EnvVars, EvalContextExt as EnvEvalContextExt};
+pub use crate::shims::io::{FileHandler, EvalContextExt as FileEvalContextExt};
```
Comment (RalfJung) [inline src/lib.rs]: `io` is extremely generic... since this is all about files and file descriptors, `file` seems like a better name to me.

File: src/shims/io.rs [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+
+        if flag != this.eval_libc_i32("O_RDONLY")? && flag != this.eval_libc_i32("O_CLOEXEC")? {
+            throw_unsup_format!("Unsupported flag {:#x}", flag);
+        }
+
+        let path_bytes = this
+            .memory()
+            .read_c_str(this.read_scalar(path_op)?.not_undef()?)?;
+        let path = std::str::from_utf8(path_bytes)
+            .map_err(|_| err_unsup_format!("{:?} is not a valid utf-8 string", path_bytes))?;
```
Comment (RalfJung) [inline src/shims/io.rs]: Here's another candidate for a factored-out helper function "get me an `OsStr` (or `OsString` probably) from memory".

File: src/shims/io.rs [hunk 17]
```diff
@@ -0,0 +1,213 @@
...
+
+pub struct FileHandler {
+    handles: HashMap<i32, FileHandle>,
+    low: i32,
+}
+
+impl Default for FileHandler {
+    fn default() -> Self {
+        FileHandler {
+            handles: Default::default(),
```
Comment (RalfJung) [inline src/shims/io.rs]: So what I do not understand here is, why is this needed? Why can't we just forward the raw FDs we got from the OS and let the program use them? What is achieved by adding this indirection?

File: src/shims/io.rs [hunk 18]
```diff
@@ -0,0 +1,213 @@
...
+        if cmd == this.eval_libc_i32("F_SETFD")? {
+            // This does not affect the file itself. Certain flags might require changing the file
+            // or the way it is accessed somehow.
+            let flag = this.read_scalar(arg_op.unwrap())?.to_i32()?;
+            // The only usage of this in stdlib at the moment is to enable the `FD_CLOEXEC` flag.
+            let fd_cloexec = this.eval_libc_i32("FD_CLOEXEC")?;
+            if let Some(FileHandle { flag: old_flag, .. }) =
+                this.machine.file_handler.handles.get_mut(&fd)
+            {
+                if flag ^ *old_flag == fd_cloexec {
```
Comment (RalfJung) [inline src/shims/io.rs]: Why `^`? That's a rare sight in flag handling code.

What I expected would have been something like `if flag & fd_cloexec != 0`, followed by `*old_flag |= fd_cloexec`.

File: src/shims/io.rs [hunk 19]
```diff
@@ -0,0 +1,213 @@
...
+                    *old_flag = flag;
+                } else {
+                    throw_unsup_format!("Unsupported arg {:#x} for `F_SETFD`", flag);
```
Comment (RalfJung) [inline src/shims/io.rs]: Note that unlike in the `open` case, it is far from obvious that this errors if e.g. any bit other than `fd_cloexec` is set... in case that's what you meant?

File: tests/run-pass/file_read.rs [hunk 20]
```diff
@@ -0,0 +1,13 @@
...
+
+use std::fs::File;
+use std::io::Read;
+
+fn main() {
+    // FIXME: create the file and delete it when `rm` is implemented.
+    let mut file = File::open("./tests/hello.txt").unwrap();
+    let mut contents = String::new();
+    file.read_to_string(&mut contents).unwrap();
+    assert_eq!("Hello, World!\n", contents);
```
Comment (RalfJung) [inline tests/run-pass/file_read.rs]: Wow, this is really cool!

File: src/shims/io.rs [hunk 21]
```diff
@@ -0,0 +1,213 @@
...
+        // Remove the file handle to avoid borrowing issues
+        this.remove_handle_and(
+            fd,
+            |mut handle, this| {
+                let bytes = handle
+                    .file
+                    .read(this.memory_mut().get_mut(buf.alloc_id)?.get_bytes_mut(
+                        tcx,
+                        buf,
+                        Size::from_bytes(count),
```
Comment (RalfJung) [inline src/shims/io.rs]: Lol, you are letting it write directly into Miri's interpreted memory? Pretty cool that that works. :D Zero-copy file access from the interpreter, hehe.

File: src/lib.rs (hunk 15 above)
Comment (RalfJung) [reply src/lib.rs]: Or maybe `fs`, to match the libstd module name.

File: src/shims/io.rs (hunk 17 above)
Comment (christianpoveda) [reply src/shims/io.rs]: The main motivation for me at least was to be able to cross-interpret code, so this should work a windows host with a linux target.

Also it should be easy to extend this to work the other way around with windows' handles

File: src/shims/io.rs (hunk 18 above)
Comment (christianpoveda) [reply src/shims/io.rs]: I thought this was a simple enough way of checking that the only change done to `flag`/`old_flag` was introducing `fd_cloexec`. For example, if you're in write mode you'll be in the following situation
`old_flag = 1`
`flag = fd_cloexec & 1`
So `flag == fd_cloexec` would be `false`.

I could something like  `flag & fd_cloexec != 0 && old_flag & fd_cloexec == 0` if you want.

File: tests/run-pass/file_read.rs (hunk 20 above)
Comment (christianpoveda) [reply tests/run-pass/file_read.rs]: You can take this as a welcome back gift :P

File: src/shims/io.rs (hunk 21 above)
Comment (christianpoveda) [reply src/shims/io.rs]: Yeah I was doing the "write into the host's memory, then write into miri's memory" way but then Oli arrived with the cooler way.

File: src/shims/io.rs (hunk 19 above)
Comment (christianpoveda) [reply src/shims/io.rs]: Yes I understand what you mean, I'm not completely happy either with this part of the code either. 

The thing is that `std::fs` sets this flag for some reason but AFAIK that flag only matters if you're executing the file and I couldn't find anything inside the standard library where using this flag actually matters.

So yeah my approach was to only allow `F_SETFD` if it was used for setting the `FD_CLOEXEC` flag. Maybe I should change the error to `"The only supported argument for `F_SETFD` is `FD_CLOEXEC`" ` or something like that.

File: src/shims/foreign_items.rs (hunk 14 above)
Comment (christianpoveda) [reply src/shims/foreign_items.rs]: work in progress here: https://github.com/rust-lang/miri/pull/983

File: src/lib.rs (hunk 15 above)
Comment (christianpoveda) [reply src/lib.rs]: https://github.com/rust-lang/miri/pull/983/commits/976c976f0948ebfb080f51c733c5bbe24c366554

File: src/shims/io.rs (hunk 18 above)
Comment (RalfJung) [reply src/shims/io.rs]: Oh I see. Yeah that should work but when you write obfuscated code like this, please explain in a comment what is happening. :)

Comment (christianpoveda) [reply src/shims/io.rs]: Yeah sorry I was in the same mood as with the escaping strings yesterday so I was cursing and reading man until it worked. Let me add the comment

File: src/shims/io.rs (hunk 19 above)
Comment (christianpoveda) [reply src/shims/io.rs]: That would work if the file is in read only mode (because someone decided `0` was a great flag value). But for example if its in append or write mode, `flags == fd_cloexec` would return false

Comment (RalfJung) [reply src/shims/io.rs]: Ah so it does have all the old flags still in there, and just adds a new one. I see.

Comment (christianpoveda) [reply src/shims/io.rs]: yes so the point is that the difference between the flags must be exactly the cloexec one.

File: src/shims/io.rs (hunk 18 above)
Comment (christianpoveda) [reply src/shims/io.rs]: https://github.com/rust-lang/miri/pull/983/commits/1c64f29811db779bb202c3450c60c227fb0234bb

File: src/shims/io.rs (hunk 19 above)
Comment (christianpoveda) [reply src/shims/io.rs]: But even then, miri is supposed to let `fcntl(fd, F_SETF, FD_CLOEXEC)` run because that's used by `std::fs` somewhere (sorry I don't remember where actually)

Comment (christianpoveda) [reply src/shims/io.rs]: pfff I'm an idiot... the easy way to do this is to check that `fcntl(fd, F_SETF, x)` is called only if `x == FD_CLOEXEC`.

Comment (RalfJung) [reply src/shims/io.rs]: > But even then, miri is supposed to let fcntl(fd, F_SETF, FD_CLOEXEC) run because that's used by std::fs somewhere (sorry I don't remember where actually)

It is used only if `F_GETFD` returns that `FD_CLOEXEC` is not set.

If you make `F_GETFD` always just return `FD_CLOEXEC`, there should be no more calls to `FD_SETFD` at all.

Comment (christianpoveda) [reply src/shims/io.rs]: It is OK to just return the actual flags for the file descriptor because the standard library always opens files with `FD_CLOEXEC` enabled: https://github.com/rust-lang/miri/pull/987

Comment (RalfJung) [reply src/shims/io.rs]: Which flags? The flags you are storing in the table are the `open` flags, not to be confused with the `fcntl` flags! libstd always sets `O_CLOEXEC` with `open`, and that *should* have the effect of `FD_CLOEXEC` being set when doing `F_GETFD`. That's why I went on about those two families of flags and even [wrote a testcase](https://github.com/rust-lang/miri/pull/962#discussion_r333901667) for you. ;)

Do not confuse the internal flags of the file with the ways these flags are controlled via syscalls. Right now, there is only one flag (on Linux) that a file internally has: `FD_CLOEXEC`. Its initial state is determined by whether `O_CLOEXEC` is set. You *could* have a `bool` for that in the table if you want, that would let us support some programs that use libc APIs directly (and not just through libstd).

All the other `O_*` flags are forgotten after the file has been opened, they are not needed any more.

Comment (christianpoveda) [reply src/shims/io.rs]: yes I just made my mind about it. I changed this here: https://github.com/rust-lang/miri/pull/987

File: src/shims/io.rs (hunk 16 above)
Comment (christianpoveda) [reply src/shims/io.rs]: https://github.com/rust-lang/miri/pull/993
